### PR TITLE
Focused Launch: use site title hook in the name step

### LINKED ISSUE #46956: Focused Launch: avoid using @wordpress/core in package hooks in order to get/set site title
Currently the `useDomainSearch()` and `useTitle()` hooks get/set the site title by accessing `core` functions. In particular:
- https://github.com/Automattic/wp-calypso/blob/204b43c25d3064be47e48edc0fa066c59a29711a/packages/launch/src/hooks/use-domain-search.ts#L18
- https://github.com/Automattic/wp-calypso/blob/204b43c25d3064be47e48edc0fa066c59a29711a/packages/launch/src/hooks/use-title.ts#L17
- https://github.com/Automattic/wp-calypso/blob/204b43c25d3064be47e48edc0fa066c59a29711a/packages/launch/src/hooks/use-title.ts#L23-L24

The new hooks introduced in https://github.com/Automattic/wp-calypso/pull/46881 should be used instead.

### PR DESCRIPTION
**Pull request by alshakero**
#### Changes proposed in this Pull Request

* Use Site data-store getter and setter to handle site title data during launch flow. The approach is also using local state to update the value when typing, similar to the behaviour in [step-by-step launch flow](https://github.com/Automattic/wp-calypso/blob/master/apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/launch-steps/name-step/index.tsx).

#### Testing instructions

1. Sandbox testingfocusedflow.wordpress.com/
2. in `apps/editing-toolkit` run `yarn dev --sync`.

#### WP-admin

3. Go to https://testingfocusedflow.wordpress.com/wp-admin/post-new.php.
4. Open console and type `wp.data.dispatch('automattic/launch' ).openFocusedLaunch()`.
5. Update the site title.
6. Title should be updated on change, and persisted onblur (check Network type to see the request).

#### Calypso
3. Run Calypso (`yarn start`).
4. Visit http://calypso.localhost:3000/post/testingfocusedflow.wordpress.com/home?flags=create/focused-launch-flow-calypso. Or the same URL but for your site.
5. Update the site title.
6. Title should be updated on change, and persisted onblur (check Network type to see the request).

Closes #46956

### CODE REVIEW

File: packages/data-stores/src/site/actions.ts [hunk 1]
```diff
@@ -127,15 +127,19 @@ export function createActions( clientCreds: WpcomClientCredentials ) {
 
 	function* setSiteTitle( siteId: number, title: string ) {
 		try {
+			// update the UI optimistically
+			yield receiveSiteTitle( siteId, title );
```
Comment (razvanpapadopol) [inline packages/data-stores/src/site/actions.ts:131]: This doesn't feel right here. Let's leave the optimistic UI update to be handled at the component level

File: packages/launch/src/focused-launch/index.tsx [hunk 2]
```diff
@@ -23,12 +23,14 @@ function noop( ...args: unknown[] ) {
...
-	const title = getSite()?.title;
-
 	const site = useSite();
 
-	const [ siteTitle, setSiteTitle ] = React.useState( title );
+	const siteTitle = useSelect( ( select ) =>
+		select( SITE_STORE ).getSiteTitle( window._currentSiteId )
+	);
+
+	const { setSiteTitle } = useDispatch( SITE_STORE );
```
Comment (razvanpapadopol) [inline packages/launch/src/focused-launch/index.tsx:32]: What do you think about creating a custom hook so the interface could look like `useState`.
Example: `const [ siteTitle, setSiteTitle ] = React.useSiteTitle();`

Once everything is working fine here, this hook could be re-used in [step-by-step launch flow](https://github.com/Automattic/wp-calypso/blob/master/apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/launch-steps/name-step/index.tsx) or in some other integrations such as the preview-with-live-title https://cloudup.com/cRg0dZffLhu

File: packages/launch/src/focused-launch/index.tsx [hunk 3]
```diff
@@ -23,12 +23,14 @@ function noop( ...args: unknown[] ) {
...
 
 const FocusedLaunch: React.FunctionComponent = () => {
-	const { getSite } = useSelect( ( select ) => select( 'core' ) ) as any;
-	const title = getSite()?.title;
-
 	const site = useSite();
 
-	const [ siteTitle, setSiteTitle ] = React.useState( title );
+	const siteTitle = useSelect( ( select ) =>
+		select( SITE_STORE ).getSiteTitle( window._currentSiteId )
```
Comment (razvanpapadopol) [inline packages/launch/src/focused-launch/index.tsx:29]: Just a note: `window._currentSiteId` won't work in [Calypso integration](https://github.com/Automattic/wp-calypso/pull/46834). Probably worth adding a comment.

**Review by razvanpapadopol — DISMISSED** [review]
> Use the `setSiteTitle` dispatcher only on submit (Launch your site) and use local state for the title.

While this approach would make sense if the site would be launched in one go, we should also consider the case when a user is doing some changes and then close the launch modal to do something else. We'll persist domain/plan selection in Launch store but site title is not only Launch related and it can be viewed/updated outside this flow. 

So let's try to follow the same pattern used in [step-by-step launch](https://github.com/Automattic/wp-calypso/blob/master/apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/launch-steps/name-step/index.tsx#L17-L28).
- initially, get site title from Site data-store (API call) and save it to local state (non persisted) that will be used at all times to display the value in launch flow.
- update only local state when typing (at `onChange`).
- save site title using Site data-store (API call) when focusing out (at `onBlur`).

**Comment by ciampo** [comment]
It'd be great if, in this PR, we could also replace a few calls to get/set the site title in the `@automattic/launch` hooks. In particular:

- https://github.com/Automattic/wp-calypso/blob/204b43c25d3064be47e48edc0fa066c59a29711a/packages/launch/src/hooks/use-domain-search.ts#L18
- https://github.com/Automattic/wp-calypso/blob/204b43c25d3064be47e48edc0fa066c59a29711a/packages/launch/src/hooks/use-title.ts#L17
- https://github.com/Automattic/wp-calypso/blob/204b43c25d3064be47e48edc0fa066c59a29711a/packages/launch/src/hooks/use-title.ts#L23-L24

File: packages/launch/src/hooks/use-title.ts [hunk 4]
```diff
@@ -2,33 +2,21 @@
...
-export function useTitle() {
-	const domain = useSelect( ( select ) => select( LAUNCH_STORE ).getSelectedDomain() );
-	const { setDomainSearch } = useDispatch( LAUNCH_STORE );
-	const [ title, setTitle ] = useEntityProp( 'root', 'site', 'title' );
-	// eslint-disable-next-line @typescript-eslint/ban-ts-comment
-	// @ts-ignore
-	const { saveEditedEntityRecord } = useDispatch( 'core' );
+export function useTitle( siteId: number ) {
+	const title = useSelect( ( select ) => select( SITE_STORE ).getSiteTitle( siteId ) );
+	const [ localStateTitle, setLocalStateTitle ] = useState( title );
```
Comment (razvanpapadopol) [inline packages/launch/src/hooks/use-title.ts:14]:
```suggestion
	const [ localStateTitle, setLocalStateTitle ] = useState( title || '' );
```
Making sure `title` is always a string here would prevent all the other checks for `undefined` (eg: [1](https://github.com/Automattic/wp-calypso/pull/46921/files#diff-2fb8ece2318e52dc186681406ed2610b204f21491260fe66331bbe668eabd625R57), [2](https://github.com/Automattic/wp-calypso/pull/46921/files#diff-8f2009196291a65edcfc3e93d032bc2aab593309be78da132e310207f0909025R46), [3](https://github.com/Automattic/wp-calypso/pull/46921/files#diff-cf261d1d8a4d4291e49ea46ec16d0a729c7dfde7be303f7fc590f6af00cbaa67R128)).

**Comment by alshakero** [comment]
Finally, this is ready for review. It drove me absolutely crazy. Turns out, chaining selectors is problematic if the original selector has a resolver. When the resolver resolves, components that use the sub-selector won't re-render.

File: packages/launch/src/focused-launch/index.tsx (hunk 2 above)
Comment (ciampo) [reply packages/launch/src/focused-launch/index.tsx]: I can see that the `useTitle` hook has been updated in `packages/launch/src/hooks/use-title.ts`

@alshakero can you confirm that this addresses the [comment](https://github.com/Automattic/wp-calypso/pull/46921#discussion_r515000875) from @razvanpapadopol ?

**Review by tjcafferkey — APPROVED** [review]
This works as described in the testing instructions and LGTM 👍🏻

**Comment by ciampo** [comment]
Not all files mentioned in #46956 were edited. I'm going to reopen that issue

**Pull request merged**